engines/io_uring.c

```diff
@@ -554,6 +554,41 @@ static struct io_u *fio_ioring_cmd_event(struct thread_data *td, int event)
 	return io_u;
 }
 
+static char *fio_ioring_cmd_errdetails(struct io_u *io_u)
+{
+	unsigned int sct = (io_u->error >> 8) & 0x7;
+	unsigned int sc = io_u->error & 0xff;
+#define MAXERRDETAIL 1024
+#define MAXMSGCHUNK 128
+	char *msg, msgchunk[MAXMSGCHUNK];
+
+	msg = calloc(1, MAXERRDETAIL);
+	strcpy(msg, "io_uring_cmd: ");
+
+	snprintf(msgchunk, MAXMSGCHUNK, "%s: ", io_u->file->file_name);
+	strlcat(msg, msgchunk, MAXERRDETAIL);
+
+	/*
+	 * Check the filename to figure out cmd_type since we don't have
+	 * thread_data here.
+	 */
+	if (strstr(io_u->file->file_name, "/dev/ng")) {
```

Comment:
This is pretty terrible, to be honest. You cannot impute type of device based on the name of the device special node. It's perfectly valid to have different names for these. This will need a rethink to be workable.

Reply:
I can think of two alternatives:

1. Change `.errdetails` to also take a `struct thread_data` argument. `td` gives you access to the ioengine options which you can use to find the `cmd_type`. You would need to bump `FIO_IOOPS_VERSION`.
2. Add a pointer to `td` to `struct io_u`.

The first choice seems like the better approach to me.

Reply:
First choice definitely the better one. And yes, it's not a problem to change the prototype of the arguments. Just remember to change ```FIO_IOOPS_VERSION``` as well.

Reply:
Updated! Thanks,